Front-end, subxt and libraries (#39)
substrate-developer-hub/docs-sandbox: source/docs/main-docs/06-build/frontend.md

Sub-section: Front-end development
Type: reference 

Application development against Substrate runtimes encompasses building user-facing interfaces such as browser and desktop applications, as well as client applications for specific hardware requirements, such as hardware wallets.
Different libraries exist to build these types of applications, depending on your needs.
This article explains the process of querying a Substrate node and using the metadata it exposes in order to provide background knowledge to help inform these decisions, or to help developers create application specific libraries.

> NOTE: Although it's technically possible to use any protocol to communicate with a Substrate node, this article assumes that the chosen interface is via a node's RPC interface using JSON-RPC.

## Metadata system 

Substrate runtimes include a sophisticated system that helps clients generate the metadata of a node by using a single RPC call.
This includes information about a pallet's storage items, transactions, events, errors and constants.

> The current version (V14) differs significantly from its predecessors as it allows clients to generate information of the types used in the runtime.

Author: Not 100% clear what your suggested change is saying here?

> The current version (V14) differs significantly from its predecessors as it allows clients to generate information of the types used in the runtime.

Reviewer: What I was getting at was: chains that use metadata V14 give developers many more options, thanks to the type information it carries. My thinking was it would be useful for many devs to know when v14 metadata went live on polkadot and kusama. I think it was deployed first to Kusama and then to Polkadot but I don't know which at block and I don't know which client version either (0.9.10? or thereabouts?).

> The current version (V14) differs significantly from its predecessors as it allows clients to generate information of the types used in the runtime.

Author: Yes this is a great idea. 👍🏻

## RPC APIs

Substrate comes with the following APIs to interact with a node:

- [`AuthorApi`](https://paritytech.github.io/substrate/master/sc_rpc/author/trait.AuthorApi.html): An API to make calls into a full node, incuding authoring extrinsics and verifying session keys.
- [`ChainApi`](https://paritytech.github.io/substrate/master/sc_rpc/chain/trait.ChainApi.html):	An API to retrieve block header and finality information.
- [`OffchainApi`](https://paritytech.github.io/substrate/master/sc_rpc/offchain/trait.OffchainApi.html): An API for	making RPC calls for off-chain workers.
- [`StateApi`](https://paritytech.github.io/substrate/master/sc_rpc/state/trait.StateApi.html): An API to query information about on-chain state such as runtime version, storage items and proofs.
- [`SystemApi`](https://paritytech.github.io/substrate/master/sc_rpc/system/trait.SystemApi.html): An API to retrieve information about network state, such as connected peers and node roles.	

## Connecting to a node

> Querying a Substrate node can either be done by using a Hypertext Transfer Protocol (HTTP) or WebSocket (WS) based JSON-RPC client, however most applications would use a WS protocol.
> The main advantage of WS is that a single connection can be reused for many messages to and from a node, whereas a typical HTTP connection allows only for a single message from, and then response to the client at a time.

Author: ```suggestion
Querying a Substrate node can either be done by using a Hypertext Transfer Protocol (HTTP) or WebSocket (WS) based JSON-RPC client.
The main advantage of WS (used in most applications) is that a single connection can be reused for many messages to and from a node, whereas a typical HTTP connection allows only for a single message from, and then response to the client at a time.
```

For this reason, if you want to subscribe to some RPC endpoint that could lead to multiple messages being returned to the client, you must use a websocket connection and not an HTTP one.
Connecting via HTTP is commonly used for fetching data in off-chain workers-learn more about that [here]().

> An alternative (and still experimental) way to connect to a JSON-RPC  is by using `Smoldot`, which allows applications to spawn their own light clients and connect directly to the exposed JSON-RPC end-point.

Author: ```suggestion
An alternative (and still experimental) way to connect to a Substrate node is by using `Substrate Connect`, which allows applications to spawn their own light clients and connect directly to the exposed JSON-RPC end-point.
These applications would rely on in-browser local memory to establish a connection with the light client. 
```

## Start building 

> Although you could use the [JSON-RPC](https://github.com/paritytech/jsonrpc) and [jsonrpsee](https://github.com/paritytech/jsonrpsee) libraries to interact with a node's RPC interface directly, Parity maintains the following libraries for interacting with Substrate runtimes:

Author: ```suggestion
Parity maintains the following libraries built on top of the [JSON-RPC API](https://github.com/paritytech/jsonrpc) for interacting with Substrate runtimes:
```

- [subxt](./libraries#subxt): provides a way to create an interface for static front-ends built for specific chains. 
- [Polkadot JS API](./libraries#polkadot-js): provides a library to build dynamic interfaces for any Substrate built blockchain.

> - [Substrate Connect](./libraries#substrate-connect): provides a library and a browser extension to build applications that connect directly with an in-browser light client created for its target chain. 
> As a library that uses the Polkadot JS API, Connect is designed for applications that need to connect to multiple chains, providing end users with a single experience when interacting with multiple chains for the same app.
> 

Author: ```suggestion
- [Substrate Connect](./libraries#substrate-connect): provides a library and a browser extension to build applications that connect directly with an in-browser light client created for its target chain. 
As a library that uses the Polkadot JS API, Connect is useful for applications that need to connect to multiple chains, providing end users with a single experience when interacting with multiple chains for the same app.

```

## Connecting to a node

Querying a Substrate node can either be done by using a Hypertext Transfer Protocol (HTTP) or WebSocket (WS) based JSON-RPC client.
The main advantage of WS (used in most applications) is that a single connection can be reused for many messages to and from a node, whereas a typical HTTP connection allows only for a single message from, and then response to the client at a time.
For this reason, if you want to subscribe to some RPC endpoint that could lead to multiple messages being returned to the client, you must use a websocket connection and not an HTTP one.
Connecting via HTTP is commonly used for fetching data in off-chain workers-learn more about that [here]().

An alternative (and still experimental) way to connect to a Substrate node is by using `Substrate Connect`, which allows applications to spawn their own light clients and connect directly to the exposed JSON-RPC end-point.
These applications would rely on in-browser local memory to establish a connection with the light client. 

## Start building 

> Parity maintains the following libraries built on top of the [JSON-RPC API](https://github.com/paritytech/jsonrpc) for interacting with Substrate runtimes:

Author: ```suggestion
Parity maintains the following libraries built on top of the [JSON-RPC API](https://github.com/paritytech/jsonrpc) for interacting with a Substrate node:
```

This includes information about a pallet's storage items, transactions, events, errors and constants.
The current metadata version (V14) differs significantly from its predecessors as it contains much richer type information. 
Polkadot uses V14 metadata starting from [runtime spec version 9110](https://polkascan.io/polkadot/runtime/9110) at [block number 7229126](https://polkadot.subscan.io/block/7229126) and Kusama from [runtime spec version 9111](https://polkascan.io/kusama/runtime/9111), at [block number 9625129](https://kusama.subscan.io/block/9625129).
This is useful to know for developers who intend to interact with runtimes that use older metadata versions.
Refer to [this document](https://gist.github.com/ascjones/0d81a4c44e84cacd9f714cd34a6de823) for a migration guide from V13 to V14.
This means that if a runtime containts a pallet with some custom type, the type information will be included as part of the metadata returned.

> This rich metadata system is made possible by both the [`frame-metadata`](https://docs.substrate.io/rustdocs/latest/frame_metadata/index.html) and [`scale-info`](https://docs.rs/scale-info/latest/scale_info/) crates which all Substrate runtimes using the latest system have.

Reviewer: This sounds a bit like these crates are part of a runtime, which is not the case, at least not in a direct sense.

`scale-info` defines the `TypeInfo` trait and a powerful derive macro used to implement it on the substrate runtime types, but this happens at build time. `frame-metadata` defines a set of types that defines the meta data itself, a sort of meta-metadata if you wish.

```suggestion
This rich metadata system is made possible by the [`scale-info`](https://docs.rs/scale-info/latest/scale_info/) crate (more details [here](TODO PROPER LINK)).
```

The general flow of how metadata is generated, exposed and used to make and receive calls from the runtime:

- Callable pallet functions, as well as types, parameters and documentation are exposed by the runtime.
- The `frame-metadata` crate describes the structure in which the information about how to communicate with the runtime will be provided. 
The information takes the form of a type registry provided by `scale-info`, as well as information about things like which pallets exist (and what the relevant types in the registry are for each pallet).
- The `scale-info` crate is used to annotate types across the runtime, and makes it possible to build a registry of runtime types. This type information is detailed enough that we can use it to find out how to correctly SCALE encode or decode some value for a given type.
- The structure described in `frame-metadata` is populated with information from the runtime, and this is then SCALE encoded and made available via the `state_getMetadata` RPC call.
- Custom RPC APIs use the metadata interface and provide methods to make calls into the runtime. 

> A SCALE codec library is required to encode and decode calls to and from these RPCs.

Reviewer: ```suggestion
A SCALE codec library is required to encode and decode calls and data to and from the API.
```

                          ┌─────────────────┐   
                          │state_getMetadata| 
                          │    RPC method   │
                          └─────────────────┘ 
                                 │   ▲
                                 ▼   │        
                            ┌─────────────┐   
                            │Type registry|  
                            └─────────────┘   
                              scale-info 
                                   ▲
                                   │                                 

>                              frame-metadata

Reviewer: A newcomer reading this might be confused thinking that these steps occur at runtime, which is not the case. `scale-info` and `frame-metadata` work their magic at compile time.

Author: Fair. I'll take this out for now. I'd like to have an architectural diagram that shows the different components we're talking about and how they're related. Just for the more visual learners.

                                   ▲
                                   │
                            ┌──────┴────────┐
                            │    Runtime    │
                            │  (n pallets)  │
                            └───────────────┘
                      
                 (Use bullet points as part above diagram)

> Every Substrate chain stores the version number of the metadata system they are using, which makes it useful for front-ends to know how to handle the metadata exposes by a certain block.

Reviewer: I'd be a little careful about using `front-ends` here as it puts funny ideas about graphical user interfaces in the head of the reader. How about just "applications"?

As previously mentioned, the latest metadata version (V14) provides a major enhancement to the metadata that a chain is able to generate.
But what if an application wants to interact with blocks that were created with an earlier version than V14?

> Well, it would require setting up a front-end interface that follows the older metadata system, whereby custom types would need to be identified and manually included as part of the front-end's code.

Reviewer: There is actually a project that can help with this, `desub` ([repo](https://github.com/paritytech/desub)). Worth linking to?

> Type information bundled in the metadata gives front-ends the ability to dynamically communicate with nodes across different chains, each of which may each expose different calls, events, types and storage.

Reviewer: "dynamical"... this is potentially true for JS/TS application but not for Rust based apps using `subxt`. A project using `subxt` is compiled against a specific metadata file and only guaranteed to work with that.

> It also allows libraries to generate almost all of the code needed to communicate with a given Substrate node, giving the possibility for libraries like `subxt` to generate front-end interfaces that are specific to a target chain.

Reviewer: ```suggestion
It also allows libraries to generate almost all of the code needed to communicate with a given Substrate node, giving the possibility for libraries like `subxt` to generate an API that is specific to a target chain.
```

With this system, any runtime can be queried for its available runtime calls, types and parameters.  

> The metadata also exposes how a type is expected to be decoded, making it easy for an external application to retrieve and process this information.

Reviewer: ```suggestion
The metadata also exposes how a type is expected to be decoded, making it easier for an external application to retrieve and process this information.
```

## Metadata format

Querying the `state_getMetadata` RPC function will return a vector of SCALE-encoded bytes which is decoded using the [`frame-metadata`](/rustdocs/latest/frame_metadata/index.html) and [`parity-scale-codec`](/rustdocs/latest/parity_scale_codec/index.html) libraries. 

> The hex blob returned by the JSON-RPC's `state_getMetadata` method depends on the metadata version, however will generally have the following structure:

Reviewer: ```suggestion
The hex blob returned by the `state_getMetadata` RPC depends on the metadata version, however will generally have the following structure:
```

("JSON-RPC" is a spec, no more no less)

- a hard-coded magic number, `0x6d657461`, which represents "meta" in plain text. 
- a 32 bit integer representing the version of the metadata format in use, for example `14` or `0x0e` in hex.
- hex encdoded type and metadata information. 
In V14, this part would contain a registry of type information (generated by the `scale-info` crate). 
In previous versions, this part contained the number of pallets followed by the metadata each pallet exposes.

> Here is a condensed version of decoded metadata for a runtime using the V14 metadata system (generated using [`subxt-cli`]()):

Reviewer: todo: fill in the link

    }
  }
]
```

As described above, the integer `1635018093` is a "magic number" that represents "meta" in plain text. 
The rest of the metadata has two sections: `pallets` and `extrinsic`. 
The `pallets` section contains information about the runtime's pallets, while the `extrinsic` section describes the version of extrinsics that the runtime is using. 
Different extrinsic versions may have different formats, especially when considering [signed extrinsics](/v3/concepts/extrinsics).

### Pallets

Here is a condensed example of a single element in the `pallets` array:

```json
{
  "name": "System",   // name of the pallet, the System pallet for example
  "storage": {        // storage entries
  },
  "calls": [          // index for this pallet's call types
  ],
  "event": [          // index for this pallet's event types
  ],
  "constants": [      // pallet constants
  ],
  "error": [         // index for this pallet's error types
  ],
  "index": 0         // the index of the pallet in the runtime
}
```

Every element contains the name of the pallet that it represents, as well as a `storage` object, `calls` array, `event` array, and `error` array.

> If `calls` or `events` are empty, they will be represented as `null` and if `constants` or `errors` are empty, they will be represented as an empty array.

Reviewer: I wasn't aware of this inconsistency, TIL. Eew!

Type indices for each item are just `u32` integers used to access the type information for that item.

> For example, the type ID for the `calls` in the System pallet is 145. 

Reviewer: Worth mentioning that these IDs are not stable over time, and not to be relied upon?

Querying the type ID will give you information about the available calls of the system pallet including the documentation for each call.
For each field, you can access type information and metadata for:

> - [Storage metadata](https://docs.substrate.io/rustdocs/latest/frame_metadata/v14/struct.PalletStorageMetadata.html): provides blockchain clients with the information that is required to query [the JSON-RPC's storage function](/rustdocs/latest/sc_rpc/state/struct.StateClient.html#method.storage) to get information for a specific storage item.

Reviewer: ```suggestion
- [Storage metadata](https://docs.substrate.io/rustdocs/latest/frame_metadata/v14/struct.PalletStorageMetadata.html): provides blockchain clients with the information that is required to query [the storage RPC](/rustdocs/latest/sc_rpc/state/struct.StateClient.html#method.storage) to get information for a specific storage item.
```

Sub-section: Front-end development
Type: reference 

> Front-end development for Substrate runtimes includes user-facing interfaces such as browser applications, desktop applications or low performance hardware environments. 

Reviewer 2: "low performance hardware environments"; should this be something like "applications for low performance hardware environments"?

Different libraries exist to build these types of interfaces, depending on your needs.
This article aims to describe:

- Types of application providers.
- How extrinsics are sent to and from an application.
- How a runtime exposes metadata that can dynamically be consumed by a front-end application for any chain.
- The process of querying metadata.
- Existing libraries for writing front-end runtime interfaces. 

## Transport providers

> Any client or front-end application requires specifying a provider to answer JSON-RPC requests made to interact with a chain.

Reviewer 2: It's not clear to me offhand what a "transport provider" is? Normally I'd assume that it is a piece of code that can be used to talk to a node (since we are communicating over RPC, an "rpc client"), and this piece of code may choose to communicate to the node via the HTTP protocol, or the WebSocket protocol. Light clients muddy the water a bit but I suppose also can act as a means to communicate with the network.

I suppose I'd want to say something like "There are multiple ways to communicate with a substrate based node; using an HTTP or WS based JSONRPC client, or a light client like Smoldot (currently in an experimental phase)". This desire might simply be because I am not properly understanding things though :)

This could be for connecting to an existing node such as `wss://rpc.polkadot.io` or connecting to a local port such as `ws://127.0.0.1:9944` when running a chain locally for example. 
Depending on the client implementation, multi-transport support is available including Websocket, HTTP and Ipc.

> Websocket connections are a more secure alternative to HTTP and are better supported among current front-end libraries.

Reviewer 2: There is no additional security for websocket connections over HTTP connections that I can think of. 

The main advantage of WS is that a single connection can be reused for many messages to and from a node, whereas a typical HTTP connection alloows only for a single message from, and then response to, the client. For this reason, if you want to subscribe to some RPC endpoint (which could lead to multiple messages being returned to the client), you must use a websocket connection and not an HTTP one.

Yet, an alternative way to specify a provider is by using `Smoldot`, which allows applications to spawn their own light clients and connect directly there (still in an experimental phase).

Several [language-specific libraries](./todo-add-link) that enable connections to these transport providers exist already.

> However, the APIs for these providers are language-agnostic, making it possible to build front-end APIs in any language.

Reviewer: I know of a python impl and have a vague memory of something in Go as well? But yeah, this concept of "providers" combined with the use of "front-end" needs a rethink.
Blockchain apps all need metadata to interact with the chain. Metadata is available from the `state_getMetadata` RPC endpoint, which works over http and ws. Any code that can speak http/ws can fetch the metadata from any node on the network. That's the easy part. The trouble starts when the app tries to decode the metadata; it's all a bag of bytes and without a full description of the content it is unuseable; any time the runtime is updated the metadata *may* change and when it does, the full description must be updated also. Polkadot.js is updated weekly (or more) and this gives third party apps a fair chance to keep working; other languages, including Rust, needs the same amount of effort (or more, as it's trickier for statically typed languages) to keep up and so far none other than javascript and polkadot.js has been a viable option.

## Sending and receiving transactions

> Specifying an endpoint to connect to will establish a connection to the specified node.

Reviewer 2: The above section seems to quite broadly look at various communication approaches. If we wish to talk JSONRPC via a nodes RPC interface, then this sentence is correct, but IPC (not familiar with how one would communicate with a node over this) or light clients are quire different I guess.

I wonder whether it's worth making it clear above that we will be talking about communicate with a node over the RPC interface, and just noting that there are other potential approaches, and then the rest can assume JSONRPC and it's all gravy?

Reviewer: IPC (aka "unix sockets") are going away, most likely in polkadot v0.9.18. (Using IPC is essentially the same as the other transports though.)

> In order to actually send and receive transactions to and from the node, all front-end APIs must implement a SCALE codec library to encode/decode RPC calls.

Reviewer: ```suggestion
In order to actually send and receive transactions to and from the node, all front-end APIs must implement a SCALE codec library to encode/decode the RPC payloads.
```

The RPC calls are JSON documents written according to the JSON-RPC spec, but it's straight JSON, no funny business there. The RPC call's parameters – the call payload – however, are almost always SCALE encoded. As a rule of thumb, anything that goes into or comes out from the runtime is SCALE encoded.

For each pallet, the metadata provides information about the storage items, transactions, events, errors and constants that are exposed by that pallet. 

> Substrate automatically generates this metadata for you and makes it available through RPC calls.

Reviewer 2: "through RPC calls" could be made more specific if you like, eg "through an RPC" call, or more specifically "through the `state_getMetadata` RPC call".

## Metadata system 

All front-end libraries that interact with Substrate runtimes are made possible by leveraging the rich metadata system provided by [`frame-metadata`](https://docs.substrate.io/rustdocs/latest/frame_metadata/index.html) and [`scale-info`](https://docs.rs/scale-info/latest/scale_info/).

                            ┌─────────────┐   
                            │RPC functions|    
                            └─────────────┘ 
                                 │   ▲
                                 ▼   │        
                            ┌─────────────┐   
                            │Type Registry|  
                            └─────────────┘   
                              scale-info 
                                   ▲
                                   │                                 
                             frame-metadata
                                   ▲
                                   │
                            ┌──────┴────────┐
                            │    Runtime    │
                            │  (n pallets)  │
                            └───────────────┘

> As shown in the above diagram:                          
> - Callable pallet functions, as well as types, parameters and documentation are exposed by the runtime.
> - The `frame-metadata` API decodes these.
> - The `scale-info` crate creates a registry of all types generated from `frame-metadata`. 
> - The types in this registry are annotated so they can be encoded and decoded to the SCALE binary format as well as information about how each type will be encoded.

Reviewer 2: I don't personally find the diagram super helpful. I guess it's sortof showing how metadata information flows from the runtime pallets to the RPC functions (though specifically just the RPC state_getMetadata  function). It's easy to look at it and assume it has something to do with general communication to/from a node, and in that case the type information is sortof a separate thing. As I type this I start wondering whether two diagrams might be useful, or just a clarification of the one.

In any case, I thought I'd have a go at clarifying the bullet points:

```suggestion
As shown in the above diagram:                          
- Callable pallet functions, as well as types, parameters and documentation are exposed by the runtime.
- The `scale-info` crate is used to annotate types across the runtime, and makes it possible to build a registry of runtime types. This type information is detailed enough that we can use it to find out how to correctly SCALE encode or decode some value for a given type.
- The `frame-metadata` crate describes the structure in which the information about how to communicate with the runtime will be provided. The information takes the form of a type registry provided by `scale-info`, as well as information about things like which pallets exist (and what the relevant types in the registry are for each pallet)
- The structure described in `frame-metadata` is populated with information from the runtime, and this is then SCALE encoded and made available via the state_getMetadata RPC call.
```

With this system, any runtime can be queried for its available runtime calls, types and parameters.  
The metadata also exposes how a type is expected to be decoded, making it easy for an external application to retrieve and process this information.

[TODO: Update examples from metadata query to V14]

## Querying metadata 

The easiest way to get the metadata is by querying the automatically-generated JSON-RPC function `state_getMetadata`. 
This will return a vector of SCALE-encoded bytes. 
You can decode this using the [`frame-metadata`](/rustdocs/latest/frame_metadata/index.html) and [`parity-scale-codec`](/rustdocs/latest/parity_scale_codec/index.html) libraries.

### Encoded metadata format

The hex blob that is returned by the JSON-RPCs `state_getMetadata` method starts with a hard-coded magic number, `0x6d657461`, which represents "meta" in plain text. 
The next piece of data (`0x0b` in the example above) represents the metadata version; decoding the hexadecimal value `0x0b` yields the

> decimal value 11, which is [the version of the Substrate metadata format](/rustdocs/latest/frame_metadata/enum.RuntimeMetadata.html) that the result encodes. 

Reviewer 2: now 14. 

Perhaps it's just worth saying something like "a 32 bit integer representing the version of the metadata format in use, for example 14".

After the metadata version, the next piece of information encoded in the result field is the number of pallets that inform the blockchain's runtime; in the example above, the hexadecimal value `0x7c` represents the decimal number 31, which is SCALE-encoded by taking its binary representation (`11111` or `0x1F` in hex), shifting it two bits to the left (`1111100`) and encoding that as hex.

> The remaining blob encodes [the metadata of each pallet](/rustdocs/latest/frame_metadata/struct.ModuleMetadata.html), which will be reviewed below as well as some [extrinsic metadata](/rustdocs/latest/frame_metadata/struct.ExtrinsicMetadata.html), which is mostly out of the scope of this document.

Reviewer 2: From V14 onwards, a big part of what is encoded in the metadata is a registry of type information (made possible by using `scale-info` to generate type information for our runtime types).

> ### Decoded metadata format

Reviewer 2: This isn't accurate any more w.r.t V14 metadata. If we keep this example I'd make it clear that it is for the older V12 metadata and future versions may (do) contain more.

Here is a condensed version of decoded metadata:

```json
{
  "magicNumber": 1635018093,
  "metadata": {
    "V12": {
      "modules": [
        {
          // ...
        },
        {
          // ...
        }
      ],
      "extrinsic": {
        "version": 4,
        "signedExtensions": [
          "CheckSpecVersion",
          "CheckTxVersion",
          "CheckGenesis",
          "CheckMortality",
          "CheckNonce",
          "CheckWeight",
          "ChargeTransactionPayment"
        ]
      }
    }
  }
}
```

As described above, the integer `1635018093` is a "magic number" that represents "meta" in plain text. 
The rest of the metadata has two sections: `modules` and `extrinsic`. 
The `modules` section contains information about the runtime's pallets, while the extrinsic section describes the version of extrinsics that the runtime is using. 
Different extrinsic versions may have different formats, especially when considering [signed extrinsics](/v3/concepts/extrinsics).

> #### Modules

Reviewer 2: These have been renamed to Pallets in at least V14 metadata, so it is now the `pallets` array.

The keys in V14 are now:

  "calls",
  "constants",
  "error",
  "event",
  "index",
  "name",
  "storage"

An example of system storage on latest polkadot:

```
{
  "name": "System",
  "storage": {
    "prefix": "System",
    "entries": [
      {
        "name": "Account",
        "modifier": "Default",
        "ty": {
          "Map": {
            "hashers": [
              "Blake2_128Concat"
            ],
            "key": 0,
            "value": 3
          }
        },
        "default": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "docs": [
          " The full account information for a particular account ID."
        ]
      },
      {
        "name": "ExtrinsicCount",
        "modifier": "Optional",
        "ty": {
          "Plain": 4
        },
        "default": [
          0
        ],
        "docs": [
          " Total extrinsics count for the current block."
        ]
      },
      {
        "name": "BlockWeight",
        "modifier": "Default",
        "ty": {
          "Plain": 7
        },
        "default": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "docs": [
          " The current weight for the block."
        ]
      },
      {
        "name": "AllExtrinsicsLen",
        "modifier": "Optional",
        "ty": {
          "Plain": 4
        },
        "default": [
          0
        ],
        "docs": [
          " Total length (in bytes) for all extrinsics put together, for the current block."
        ]
      },
      {
        "name": "BlockHash",
        "modifier": "Default",
        "ty": {
          "Map": {
            "hashers": [
              "Twox64Concat"
            ],
            "key": 4,
            "value": 9
          }
        },
        "default": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "docs": [
          " Map of block numbers to block hashes."
        ]
      },
      {
        "name": "ExtrinsicData",
        "modifier": "Default",
        "ty": {
          "Map": {
            "hashers": [
              "Twox64Concat"
            ],
            "key": 4,
            "value": 10
          }
        },
        "default": [
          0
        ],
        "docs": [
          " Extrinsics data for the current block (maps an extrinsic's index to its data)."
        ]
      },
      {
        "name": "Number",
        "modifier": "Default",
        "ty": {
          "Plain": 4
        },
        "default": [
          0,
          0,
          0,
          0
        ],
        "docs": [
          " The current block number being processed. Set by `execute_block`."
        ]
      },
      {
        "name": "ParentHash",
        "modifier": "Default",
        "ty": {
          "Plain": 9
        },
        "default": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "docs": [
          " Hash of the previous block."
        ]
      },
      {
        "name": "Digest",
        "modifier": "Default",
        "ty": {
          "Plain": 11
        },
        "default": [
          0
        ],
        "docs": [
          " Digest of the current block, also part of the block header."
        ]
      },
      {
        "name": "Events",
        "modifier": "Default",
        "ty": {
          "Plain": 15
        },
        "default": [
          0
        ],
        "docs": [
          " Events deposited for the current block.",
          "",
          " NOTE: This storage item is explicitly unbounded since it is never intended to be read",
          " from within the runtime."
        ]
      },
      {
        "name": "EventCount",
        "modifier": "Default",
        "ty": {
          "Plain": 4
        },
        "default": [
          0,
          0,
          0,
          0
        ],
        "docs": [
          " The number of events in the `Events<T>` list."
        ]
      },
      {
        "name": "EventTopics",
        "modifier": "Default",
        "ty": {
          "Map": {
            "hashers": [
              "Blake2_128Concat"
            ],
            "key": 9,
            "value": 142
          }
        },
        "default": [
          0
        ],
        "docs": [
          " Mapping between a topic (represented by T::Hash) and a vector of indexes",
          " of events in the `<Events<T>>` list.",
          "",
          " All topic vectors have deterministic storage locations depending on the topic. This",
          " allows light-clients to leverage the changes trie storage tracking mechanism and",
          " in case of changes fetch the list of events of interest.",
          "",
          " The value has the type `(T::BlockNumber, EventIndex)` because if we used only just",
          " the `EventIndex` then in case if the topic has the same contents on the next block",
          " no notification will be triggered thus the event might be lost."
        ]
      },
      {
        "name": "LastRuntimeUpgrade",
        "modifier": "Optional",
        "ty": {
          "Plain": 143
        },
        "default": [
          0
        ],
        "docs": [
          " Stores the `spec_version` and `spec_name` of when the last runtime upgrade happened."
        ]
      },
      {
        "name": "UpgradedToU32RefCount",
        "modifier": "Default",
        "ty": {
          "Plain": 59
        },
        "default": [
          0
        ],
        "docs": [
          " True if we have upgraded so that `type RefCount` is `u32`. False (default) if not."
        ]
      },
      {
        "name": "UpgradedToTripleRefCount",
        "modifier": "Default",
        "ty": {
          "Plain": 59
        },
        "default": [
          0
        ],
        "docs": [
          " True if we have upgraded so that AccountInfo contains three types of `RefCount`. False",
          " (default) if not."
        ]
      },
      {
        "name": "ExecutionPhase",
        "modifier": "Optional",
        "ty": {
          "Plain": 140
        },
        "default": [
          0
        ],
        "docs": [
          " The execution phase of the block."
        ]
      }
    ]
  },
  "calls": {
    "ty": 145
  },
  "event": {
    "ty": 18
  },
  "constants": [
    {
      "name": "BlockWeights",
      "ty": 150,
      "value": [
        0,
        242,
        5,
        42,
        1,
        0,
        0,
        0,
        0,
        32,
        74,
        169,
        209,
        1,
        0,
        0,
        64,
        89,
        115,
        7,
        0,
        0,
        0,
        0,
        1,
        192,
        118,
        108,
        143,
        88,
        1,
        0,
        0,
        1,
        0,
        152,
        247,
        62,
        93,
        1,
        0,
        0,
        1,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        64,
        89,
        115,
        7,
        0,
        0,
        0,
        0,
        1,
        192,
        254,
        190,
        249,
        204,
        1,
        0,
        0,
        1,
        0,
        32,
        74,
        169,
        209,
        1,
        0,
        0,
        1,
        0,
        136,
        82,
        106,
        116,
        0,
        0,
        0,
        64,
        89,
        115,
        7,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "docs": [
        " Block & extrinsics weights: base values and limits."
      ]
    },
    {
      "name": "BlockLength",
      "ty": 154,
      "value": [
        0,
        0,
        60,
        0,
        0,
        0,
        80,
        0,
        0,
        0,
        80,
        0
      ],
      "docs": [
        " The maximum length of a block (in bytes)."
      ]
    },
    {
      "name": "BlockHashCount",
      "ty": 4,
      "value": [
        96,
        9,
        0,
        0
      ],
      "docs": [
        " Maximum number of block number to block hash mappings to keep (oldest pruned first)."
      ]
    },
    {
      "name": "DbWeight",
      "ty": 156,
      "value": [
        64,
        120,
        125,
        1,
        0,
        0,
        0,
        0,
        0,
        225,
        245,
        5,
        0,
        0,
        0,
        0
      ],
      "docs": [
        " The weight of runtime database operations the runtime can invoke."
      ]
    },
    {
      "name": "Version",
      "ty": 157,
      "value": [
        32,
        112,
        111,
        108,
        107,
        97,
        100,
        111,
        116,
        60,
        112,
        97,
        114,
        105,
        116,
        121,
        45,
        112,
        111,
        108,
        107,
        97,
        100,
        111,
        116,
        0,
        0,
        0,
        0,
        180,
        35,
        0,
        0,
        0,
        0,
        0,
        0,
        56,
        223,
        106,
        203,
        104,
        153,
        7,
        96,
        155,
        4,
        0,
        0,
        0,
        55,
        227,
        151,
        252,
        124,
        145,
        245,
        228,
        1,
        0,
        0,
        0,
        64,
        254,
        58,
        212,
        1,
        248,
        149,
        154,
        5,
        0,
        0,
        0,
        210,
        188,
        152,
        151,
        238,
        208,
        143,
        21,
        3,
        0,
        0,
        0,
        247,
        139,
        39,
        139,
        229,
        63,
        69,
        76,
        2,
        0,
        0,
        0,
        175,
        44,
        2,
        151,
        162,
        62,
        109,
        61,
        2,
        0,
        0,
        0,
        73,
        234,
        175,
        27,
        84,
        138,
        12,
        176,
        1,
        0,
        0,
        0,
        145,
        213,
        223,
        24,
        176,
        210,
        207,
        88,
        1,
        0,
        0,
        0,
        237,
        153,
        197,
        172,
        178,
        94,
        237,
        245,
        3,
        0,
        0,
        0,
        203,
        202,
        37,
        227,
        159,
        20,
        35,
        135,
        2,
        0,
        0,
        0,
        104,
        122,
        212,
        74,
        211,
        127,
        3,
        194,
        1,
        0,
        0,
        0,
        171,
        60,
        5,
        114,
        41,
        31,
        235,
        139,
        1,
        0,
        0,
        0,
        188,
        157,
        137,
        144,
        79,
        91,
        146,
        63,
        1,
        0,
        0,
        0,
        55,
        200,
        187,
        19,
        80,
        169,
        162,
        168,
        1,
        0,
        0,
        0,
        9,
        0,
        0,
        0,
        0
      ],
      "docs": [
        " Get the chain's current version."
      ]
    },
    {
      "name": "SS58Prefix",
      "ty": 74,
      "value": [
        0,
        0
      ],
      "docs": [
        " The designated SS85 prefix of this chain.",
        "",
        " This replaces the \"ss58Format\" property declared in the chain spec. Reason is",
        " that the runtime should know about the prefix in order to make use of it as",
        " an identifier of the chain."
      ]
    }
  ],
  "error": {
    "ty": 161
  },
  "index": 0
}
```

Reviewer 2: Building on the above, we can see that the type of the "calls" in this System pallet is 145.

If I dig into the JSON decoded metadata that I get back for latest polkadot with the JQ query `jq '.[1].V14.types.types | .[145]'`, I can see the information for that type:

```
{
  "id": 145,
  "type": {
    "path": [
      "frame_system",
      "pallet",
      "Call"
    ],
    "params": [
      {
        "name": "T",
        "type": null
      }
    ],
    "def": {
      "variant": {
        "variants": [
          {
            "name": "fill_block",
            "fields": [
              {
                "name": "ratio",
                "type": 146,
                "typeName": "Perbill"
              }
            ],
            "index": 0,
            "docs": [
              "A dispatch that will fill the block weight up to the given ratio."
            ]
          },
          {
            "name": "remark",
            "fields": [
              {
                "name": "remark",
                "type": 10,
                "typeName": "Vec<u8>"
              }
            ],
            "index": 1,
            "docs": [
              "Make some on-chain remark.",
              "",
              "# <weight>",
              "- `O(1)`",
              "# </weight>"
            ]
          },
          {
            "name": "set_heap_pages",
            "fields": [
              {
                "name": "pages",
                "type": 8,
                "typeName": "u64"
              }
            ],
            "index": 2,
            "docs": [
              "Set the number of pages in the WebAssembly environment's heap."
            ]
          },
          {
            "name": "set_code",
            "fields": [
              {
                "name": "code",
                "type": 10,
                "typeName": "Vec<u8>"
              }
            ],
            "index": 3,
            "docs": [
              "Set the new runtime code.",
              "",
              "# <weight>",
              "- `O(C + S)` where `C` length of `code` and `S` complexity of `can_set_code`",
              "- 1 call to `can_set_code`: `O(S)` (calls `sp_io::misc::runtime_version` which is",
              "  expensive).",
              "- 1 storage write (codec `O(C)`).",
              "- 1 digest item.",
              "- 1 event.",
              "The weight of this function is dependent on the runtime, but generally this is very",
              "expensive. We will treat this as a full block.",
              "# </weight>"
            ]
          },
          {
            "name": "set_code_without_checks",
            "fields": [
              {
                "name": "code",
                "type": 10,
                "typeName": "Vec<u8>"
              }
            ],
            "index": 4,
            "docs": [
              "Set the new runtime code without doing any checks of the given `code`.",
              "",
              "# <weight>",
              "- `O(C)` where `C` length of `code`",
              "- 1 storage write (codec `O(C)`).",
              "- 1 digest item.",
              "- 1 event.",
              "The weight of this function is dependent on the runtime. We will treat this as a full",
              "block. # </weight>"
            ]
          },
          {
            "name": "set_storage",
            "fields": [
              {
                "name": "items",
                "type": 147,
                "typeName": "Vec<KeyValue>"
              }
            ],
            "index": 5,
            "docs": [
              "Set some items of storage."
            ]
          },
          {
            "name": "kill_storage",
            "fields": [
              {
                "name": "keys",
                "type": 149,
                "typeName": "Vec<Key>"
              }
            ],
            "index": 6,
            "docs": [
              "Kill some items from storage."
            ]
          },
          {
            "name": "kill_prefix",
            "fields": [
              {
                "name": "prefix",
                "type": 10,
                "typeName": "Key"
              },
              {
                "name": "subkeys",
                "type": 4,
                "typeName": "u32"
              }
            ],
            "index": 7,
            "docs": [
              "Kill all storage items with a key that starts with the given prefix.",
              "",
              "**NOTE:** We rely on the Root origin to provide us the number of subkeys under",
              "the prefix we are removing to accurately calculate the weight of this function."
            ]
          },
          {
            "name": "remark_with_event",
            "fields": [
              {
                "name": "remark",
                "type": 10,
                "typeName": "Vec<u8>"
              }
            ],
            "index": 8,
            "docs": [
              "Make some on-chain remark and emit event.",
              "",
              "# <weight>",
              "- `O(b)` where b is the length of the remark.",
              "- 1 event.",
              "# </weight>"
            ]
          }
        ]
      }
    },
    "docs": [
      "Contains one variant per dispatchable that can be called by an extrinsic."
    ]
  }
}
```

The possible calls you can make into the system pallet are represented by this variant type, where each variant is a call, for instance "fill_block". Each "field" of the variant is an argument that you need to provide when making that call, and the type ID of each argument can also be looked up in the type registry to find more info about each parameter.

Here is a condensed example of a single element in the `modules` array:

```json
{
  "name": "System",
  "storage": {
    // ..
  },
  "calls": [
    // ..
  ],
  "events": [
    // ..
  ],
  "constants": [
    // ..
  ],
  "errors": [
    // ..
  ],
  "index": 0
}
```

Every element contains the name of the pallet that it represents, as well as a `storage` object, `calls` array, `events` array, and `errors` array.

<Message
  type={`gray`}
  title={`Note`}
  text={`If \`calls\` or \`events\` are empty, they will be represented as \`null\`; if \`constants\` or
 \`errors\` are empty, they will be represented as an empty array.`}
/>

##### Storage

Here is a condensed example of a single element in the `modules` array that highlights metadata about the module's storage:

> ```json

Reviewer 2: See the above; this is no longer the correct shape with V14 metadata

      }
    ]
  },
  "calls": [
    /*...*/
  ],
  "events": [
    /*...*/
  ],
  "constants": [
    /*...*/
  ],
  "errors": [
    /*...*/
  ],
  "index": 0
}
```

Every storage item that is defined in a pallet will have a corresponding metadata entry.
Metadata entries like these are generated from [macros](/v3/runtime/macros) using associated types from the [`frame-system`](/rustdocs/latest/frame_system/pallet/trait.Config.html) crate. 
For example:

```rust
#[pallet::config]
pub trait Config: frame_system::Config {
	#[pallet::constant]
	type Foo: Get<u32>;
}
```

[Storage metadata](/rustdocs/latest/frame_metadata/struct.StorageMetadata.html) provides blockchain clients with the information that is required to query [the JSON-RPC's storage function](/rustdocs/latest/sc_rpc/state/struct.StateClient.html#method.storage) to get information for a specific storage item.

##### Dispatchable calls

Metadata for dispatchable calls includes information about the runtime calls are defined by the `#[pallet]` macro. For each call, the metadata includes:

- `name`: Name of the function in the pallet.
- `args`: Arguments in function definition. Includes the name and type of each argument.

> - `documentation`: Documentation of the function.

Reviewer 2: Note the above metadata examples I gave; a given call now looks like:

```
          {
            "name": "fill_block",
            "fields": [
              {
                "name": "ratio",
                "type": 146,
                "typeName": "Perbill"
              }
            ],
            "index": 0,
            "docs": [
              "A dispatch that will fill the block weight up to the given ratio."
            ]
          },
```

Name of call, each field is a parameter to pass to it (and we can look up the type info for them too) and docs.

      " and the new runtime."
    ]
  },
  // ...
]
```

These are errors that could occur during the submission or execution of an extrinsic. 
In this case, the FRAME System pallet is declaring that it may raise the the [`InvalidSpecName` error](/rustdocs/latest/frame_system/pallet/enum.Error.html#variant.InvalidSpecName).

## Libraries 

Different libraries exist for building frontend interfaces for Substrate-based chains and interacting with Substrate runtimes.
- [`subxt`](#subxt) provides a way to fetch the metadata and decode them for you. 
- [Polkadot JS API](#polkadot-js) provides multiple interfaces to interact with a Substrate blockchain.
- [Substrate Connect](#substrate-connect) provides a library to build applications that connect directly with a target chain.

To use the RPC more directly, you can use the [JSONRPC](https://github.com/paritytech/jsonrpc) and [jsonrpsee](https://github.com/paritytech/jsonrpsee) Rust libraries.
### subxt 

`subxt` is a Rust library for submitting extrinsics to a Substrate node.

It provides a way for developers to write user facing client applications for any Substrate chain without the burden of maintaining a list of types. 

> It's also a good solution for lower level applications, such as non-browser graphic user interfaces or building an application specific CLI.

Reviewer 2: Crucially, `subxt` generates an interface to a _specific_ substrate node based on the current metadata that it returns. This allows for typesafe communication between the node and the code using this generated interface.

As well as submitting extrinsics, `subxt` also allows one to access storage entries, decode blocks and decode the events contained within the blocks.

                              scale-info 
                                   ▲
                                   │                                 
                             frame-metadata
                                   ▲
                                   │
                            ┌──────┴────────┐
                            │    Runtime    │
                            │  (n pallets)  │
                            └───────────────┘

As shown in the above diagram:                          
- Callable pallet functions, as well as types, parameters and documentation are exposed by the runtime.
- The `frame-metadata` API decodes these.
- The `scale-info` crate creates a registry of all types generated from `frame-metadata`. 
- The types in this registry are annotated so they can be encoded and decoded to the SCALE binary format as well as information about how each type will be encoded.

With this system, any runtime can be queried for its available runtime calls, types and parameters.  
The metadata also exposes how a type is expected to be decoded, making it easy for an external application to retrieve any of the information a runtime exposes. 

> [add content from "c-metadata"]

Author: ```suggestion
[TODO: add content from "c-metadata"]
```

Sub-section: Front-end development
Type: reference 

Application development against Substrate runtimes encompasses building user-facing interfaces such as browser and desktop applications, as well as client applications for specific hardware requirements, such as hardware wallets.
Different libraries exist to build these types of applications, depending on your needs.
This article explains the process of querying a Substrate node and using the metadata it exposes in order to provide background knowledge to help inform these decisions, or to help developers create application specific libraries.

## Metadata system 

Substrate nodes provide an RPC call, `state_getMetadata`, that returns a complete description of all the types in the current runtime. Client applications use the metadata to interact with the node, to parse responses and format message payloads sent to the node.
This includes information about a pallet's storage items, transactions, events, errors and constants.

> The current metadata version (V14) differs significantly from its predecessors as it contains much richer type information. 

Reviewer 3: I'm not sure that this is really all that helpful here.
Maybe instead something about how the information stored in the metadata depends on the version of the Substrate node you have compiled.
For example, the current version of the metadata includes information for custom types.
```suggestion
The information stored in the metadata depends on the version of the Substrate node you have compiled.
For example, the current version of the metadata includes information for custom types.
```

Reviewer: There's a significant difference to what builders can do with V14 metadata vs previous versions. Entire classes of applications were previously nigh impossible to build (e.g. anything using Rust) so while this edit is trivially true, it also leaves out this important piece of information.

> This means that if a runtime containts a pallet with some custom type, the type information will be included as part of the metadata returned.

Reviewer 3: ```suggestion
If a runtime includes a pallet with a custom type, the type information is included as part of the metadata returned.
```

> Polkadot uses V14 metadata starting from [runtime spec version 9110](https://polkascan.io/polkadot/runtime/9110) at [block number 7229126](https://polkadot.subscan.io/block/7229126) and Kusama from [runtime spec version 9111](https://polkascan.io/kusama/runtime/9111), at [block number 9625129](https://kusama.subscan.io/block/9625129).

Reviewer 3: These three sentences seem a little weird to me. It might be useful to explain how you can look up the version of the metadata your blockchain uses, and something like:
If you need to interact with a runtime that uses an older version of the metadata, see [Migrating between metadata versions]().

Author: I'll put a note to fix this once its in the `main-md` docs.

This is useful to know for developers who intend to interact with runtimes that use older metadata versions.
Refer to [this document](https://gist.github.com/ascjones/0d81a4c44e84cacd9f714cd34a6de823) for a migration guide from V13 to V14.

> This rich metadata system is made possible by the [`scale-info`](https://docs.rs/scale-info/latest/scale_info/) crate, used to get runtime type information at build time.

Reviewer 3: ```suggestion
The current metadata schema uses the [`scale-info`](https://docs.rs/scale-info/latest/scale_info/) crate to get type information for the pallets in the runtime when you compile a node.
```

Reviewer: There is no "schema" for the type information unless the source code of the `scale_info::TypeInfo` trait and it's serialization to JSON. I agree that "system" is perhaps not the best possible term, but I do see where @sacha-l is coming from.

> In order to actually send and receive transactions to and from the node, front-end APIs must implement a [SCALE codec library](./libraries#SCALE-Codec) to encode and decode RPC payloads.

Reviewer 3: ```suggestion
The current implementation of the metadata requires front-end APIs to use the [SCALE codec library](./libraries#SCALE-Codec) to encode and decode RPC payloads to send and receive transactions.
```

Reviewer: SCALE is the codec; which library to use to encode/decode in SCALE depends on the app (front-end apps would use a JS library, likely `polkadot.js`, rust apps use the `parity-scale-codec` crate, python apps `python-scale-codec` etc), so I think using "a library" is more correct here.

Also: talking about "front-end" apps/apis is not correct: the scoper is wider than browser-based apps.

> The general flow of how metadata is generated, exposed and used to make and receive calls from the runtime:

Reviewer 3: ```suggestion
The following steps summarize how metadata is generated, exposed, and used to make and receive calls from the runtime:
```

Reviewer 3: This is as far as I've gotten :wink:
